### DISCUSSION TITLE
How do you make a package do things

### DISCUSSION THREAD

**Discussion by zeventhal** [livebud/bud#107] [Q&A] [answered]
I read the docs / skimmed and I want to contribute but I just really don't understand how the actual bud project is structured, or how plugins are supposed to be written. Some help would be great because this project is really cool and I want to add to it and make it a full featured full stack framework similar to rails.

**Comment by matthewmueller** [comment, marked as answer]
Hey @zeventhal! Thanks for your interest in the project! I definitely need to get more architecture posts up, but for now I recommend having a look at how the compiler works in here: https://github.com/livebud/bud/discussions/18

Then hop over to the [Contributing Guide](https://github.com/livebud/bud/tree/main/contributing), clone the project and just start playing around. It all starts from the project's root `main.go` file, so you can start from there and start exploring.

Additional details:
- `internal/`: contains internal packages that don't need to be access by the generated project.
- `packages/`: generic packages that aren't really tied to Bud itself and could be their own Go Module.
- `runtime/` packages needed by the generated code.
- `livebud/` runtime equivalent on the Javascript side.

_(Warning: The above details may change as the project progresses. I'm about 70% happy with the directory structure right now)_
